### PR TITLE
Feature/253 introduce community add ons

### LINKED ISSUE #253: Introduce community plugin management framework
# Product User Story

**Problem**
Currently, community admins do not have a consistent and easy way to personalize their community experience. For example, the GitHub and Kanban  components always display, and an admin must manage them within the tab and cannot remove them. 

**User Story**
As an admin, I need a consistent and intuitive method to add, configure, and remove add-ons for my community via the community settings page.

**Acceptance Criteria**
- By default, any newly created communities will not display any of the available add-ons. So only the main feed and teams tab will be enabled.
- Admins have a consistent framework for adding, configuring, and deleting add-ons all through the community settings page.
- Adding: Admins can add add-ons through a "Add add-on" button and select from a dropdown menu of available add-ons. An admin can add multiple instances of the same add-on. By default, the naming of each tab will be the add-on name.
- Configuring: Once a new add-on is added, admins can click edit to customize the add-on’s tab names and specific fields, and click save to apply their changes. 
  - Previewing: Admins should have a way to quickly preview their tab in the settings page. Perhaps we add a toggle here similar to GitHub's "Write/Preview" interface to avoid crowding the interface for more complex plugin configurations?
- Deletion: Admins can delete add-ons by accessing the edit option and selecting delete, removing the respective tab.
- Migration: Existing add-on configurations are preserved, and communities can manually migrate at their convenience without breaking functionality.
- Tab names
   - Limit # of characters in tab to X
   - Limit # of widgets: Only allow up to ~# number of widgets to fit under the navigation

----------

# Old Technical Notes

**Is your feature request related to a problem? Please describe.**
Currently a DevHUB community only allows built-in pages/components like Telegram and Github, limiting the potential for unique functionalities tailored to the needs of each community. Also the current way of adding Telegram and Github is by registering handles/URLs in the central configuration section, which does not give the experience of a "pluggable" component with its own configuration.

**Describe the solution you'd like**
I would like Telegram, Github and BOS widgets in general to be components that can be added from the configuration page, and also have their own configuration sections. A user experience with a button for adding a "widget" like the following screenshot:

image

When clicking "Add widget" you should be able to select from a drop-down menu different alternatives of built-in (like Telegram and Github) as well of community provided widgets: 

image

Finally when a custom widget like the "Music tracker" is added, there's a new configuration section in the community admin:

image

and a new tab with the widget in the community user page:

image

**Example use cases**

An example use case is creating a tab with custom activity feed, so let’s say you pick the Feed widget from the as a component to add, and then it has a configuration called filter, you out filter of tag let’s say blogs and the community name, so now you have a feed of blogs for this community. The tab is renamed to "Blog".

You could even add the feed widget again to another tab, and now with a different filter configuration. For example now you want all the posts from `dev-dao`, and so you configure the widget with the filter `dev-dao`. Also you rename that tab with the name `Dev DAO`.

So now you have the Feed widget present twice, but with different settings.

**Additional context**
- In order to preserve existing configuration for Github and Telegram, the change must not break communities that have not migrated to the new approach
- Instead of migrating every community when this new feature is merged, rather communities should be able to migrate manually when ready, this means that the current approach also need to be available, and community upgrade should be on demand.
- Migration should be the responsibility of new Telegram/Github/Board widgets respectively which are developed separately ( separate tickets to be provided )
- Configuration for each community component should be stored in the DevHUB contract
- Widgets should be regular BOS widgets
- For the MVP, the list of widgets that can be added should be hardcoded into the DevHUB UI, later a widget registry should be added ( separate ticket to be provided )
- A community component needs to support two view modes: one for the component to appear in the new tab, and one for configuration section
- It should be possible to alter the tab name
- It should be possible to delete a widget

### PR DESCRIPTION
**Pull request by Tguntenaar**
_Resolves [#253](https://github.com/near/neardevhub-widgets/issues/253)_

The following structs have been added:
schema

- [x] Add structs
- [x] Implement migration
- [x] Test migration manually
- [x] Add basic CRUD operations
- [x] Permissions, only community admins and moderators can create, update, delete add-ons to communities
- [x] Only moderators can create, update, delete add-ons to available add-ons
- [x] Write tests for operations
- [x] Edit test for migration

Acceptence Criteria (not all AC apply to the smart contract side)
- [x] By default, any newly created communities will not display any of the available add-ons. So only the main feed and teams tab will be enabled.
- [x] Admins have a consistent framework for adding, configuring, and deleting add-ons all through the community settings page.
- [x] Adding: Admins can add add-ons through a "Add add-on" button and select from a dropdown menu of available add-ons. An admin can add multiple instances of the same add-on. By default, the naming of each tab will be the add-on name.
- [x] Configuring: Once a new add-on is added, admins can click edit to customize the add-on’s tab names and specific fields, and click save to apply their changes.
- [x] Previewing: Admins should have a way to quickly preview their tab in the settings page. Perhaps we add a toggle here similar to GitHub's "Write/Preview" interface to avoid crowding the interface for more complex plugin configurations?
- [x] Deletion: Admins can delete add-ons by accessing the edit option and selecting delete, removing the respective tab.
Migration: Existing add-on configurations are preserved, and communities can manually migrate at their convenience without breaking functionality.

I tested the migration with the following transactions:
[From V7 to V8](https://explorer.testnet.near.org/transactions/ndCDxCZm7AzY1tS33Lrn1G9zb2LtLSupj9gYiKZD4ue)
[V8 to 'done'](https://explorer.testnet.near.org/transactions/8jp8j9sP2XxXkNgmZoz6ggkdc1jKWV3K5DHGz5SyPBLQ)

### CODE REVIEW

**Comment by Tguntenaar** [comment]
@ailisp , I discussed with elliot that the function to delete available addons were out of scope. Removing an addon from a community tab is possible. Let me know what you think!

**Comment by Tguntenaar** [comment]
Hi @frol, since @ailisp is OOO. Could you review this PR? It's vital to continue the [Community Plugins MVP](https://github.com/near/neardevhub-widgets/milestone/10) milestone.

**Comment by petersalomonsen** [comment]
would it be better to use a `LookupMap` for the community plugin configuration data? That way it would be possible to look up the config for the plugin in view, and also not make the community config object potentially very large.

**Comment by elliotBraem** [comment]
@petersalomonsen 

Can you demonstrate? Is this is to replace parameters JSON string?

**Comment by petersalomonsen** [comment]
yeah I was thinking that `pub addon_list: Vec<CommunityAddOnConfig>` could just be a list of strings, an those strings represent the keys in a `LookupMap<String,CommunityAddonConfig>` so that each config has a separate entry in the NEAR key/value storage. I believe this will scale better with multiple addons, or addons with larger config objects.

File: src/lib.rs [hunk 1]
```diff
@@ -413,6 +416,126 @@ impl Contract {
...
+        }
+        if self.get_addon(addon_config.addon_id.to_owned()).is_none() {
+            panic!("No add-on exists with this id");
+        }
+
+        community.add_addon(addon_config);
+        self.communities.insert(&community_handle, &community);
+    }
+
+    pub fn edit_community_addon(
```
Comment (elliotBraem) [inline src/lib.rs]: "update_community_addon" may be a better name

Comment (elliotBraem) [reply src/lib.rs]: Will removing and then creating modify the addon's id? And so references elsewhere will break once edited?

(see Peter's latest reply for a possible solution)

File: src/lib.rs [hunk 2]
```diff
@@ -413,6 +416,126 @@ impl Contract {
...
+            format!("Community with handle `{}` does not exist", community_handle).as_str(),
+        );
+
+        if !self.has_moderator(env::predecessor_account_id())
+            && env::predecessor_account_id() != env::current_account_id()
+            && !community.admins.contains(&env::predecessor_account_id())
+        {
+            panic!("Only moderators and community admins can remove add-ons from a community");
+        }
+        community.remove_addon(config_id);
```
Comment (elliotBraem) [inline src/lib.rs]: Am having issues with this.

```javascript
const handleDeleteAddon = (addon_id) => {
  DevHub.remove_community_addon({ handle, config_id: addon_id });
};
```

This transaction did not do anything:
https://explorer.testnet.near.org/transactions/AWu7ubj98z51mB2ejNtEYCC49XCetzcXqHYYFGPz9w9H

File: src/lib.rs [hunk 3]
```diff
@@ -413,6 +416,126 @@ impl Contract {
...
+        community
+            .addon_list
+            .iter()
+            .map(|addon_config| {
+                self.get_addon(addon_config.addon_id.clone()).expect("add-on missing")
+            })
+            .collect()
+    }
+
+    pub fn add_community_addon(
```
Comment (elliotBraem) [inline src/lib.rs]: Could we have this method generate the addon config id?

See this transaction panic because config_id is not set: https://explorer.testnet.near.org/transactions/8THKLRov1FAKbj7iSu4vhCX3NniBzqvsh3a221PqZqXH

Could be considered with Peter's suggestion

File: src/lib.rs (hunk 1 above)
Comment (Tguntenaar) [reply src/lib.rs]: You could in theory update the `addon_id` in CommunityAddOnConfig. It would point to a different add on. However not the `config_id` (because remove_ would panic) which wouldn't break the reference made by peters comment.

I updated the name of the function to: update_community_addon

File: src/lib.rs (hunk 2 above)
Comment (Tguntenaar) [reply src/lib.rs]: Hmm that's weird it does work for me [here](https://explorer.testnet.near.org/transactions/CPCNimNFFDmgBiA4kYFz3uHhRsqwNqjGNcPNbJXs8iAX)
I will look into this.

**Comment by Tguntenaar** [comment]
I updated the contract to use a LookupMap instead. Thanks @petersalomonsen for the review. New contract deployed here:
https://explorer.testnet.near.org/transactions/DEu1TwCZM2fBbpqWZxZBL62ik6et7bCC9DeRbaMD4sdh

File: src/lib.rs (hunk 3 above)
Comment (Tguntenaar) [reply src/lib.rs]: I think it would be better to generate or choose such an id from the front end side.

**Comment by Tguntenaar** [comment]
@ailisp could you review these contract changes?

File: src/community/mod.rs [hunk 4]
```diff
@@ -45,6 +49,54 @@ pub struct WikiPage {
...
 }
 
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone)]
+#[serde(crate = "near_sdk::serde")]
+pub struct CommunityAddOnConfig {
+    pub config_id: CommunityAddOnConfigId,
+    pub addon_id: CommunityAddOnId,
+    pub parameters: String,
+    pub enabled: bool,
+    pub name: String,
```
Comment (frol) [inline src/community/mod.rs]: What is `name` for?

File: src/community/mod.rs [hunk 5]
```diff
@@ -45,6 +49,54 @@ pub struct WikiPage {
...
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, PartialEq, Debug)]
+#[serde(crate = "near_sdk::serde")]
+pub struct CommunityAddOn {
+    pub id: CommunityAddOnId,
+    pub title: String,
+    pub description: String,
+    pub viewer: String,
+    pub configurator: String,
```
Comment (frol) [inline src/community/mod.rs]: What are these? (please, add more comments in the code or make the naming self-descriptive)

File: src/community/mod.rs [hunk 6]
```diff
@@ -110,6 +163,14 @@ impl Community {
         }
     }
 
+    pub fn add_addon(&mut self, addon_config: CommunityAddOnConfigId) {
+        self.addon_list.push(addon_config);
+    }
+
+    pub fn remove_addon(&mut self, config_id: String) {
+        self.addon_list.retain(|config| config != &config_id);
+    }
```
Comment (frol) [inline src/community/mod.rs]: How do you envision the layout of addons on the community page (on UI)? If we display them as tabs, there should be a way to order them, so instead of removing them, we may have `set_addons` where the whole list of addon configs will be expected.

File: src/community/mod.rs [hunk 7]
```diff
@@ -67,6 +119,7 @@ pub struct Community {
     pub wiki1: Option<WikiPage>,
     pub wiki2: Option<WikiPage>,
     pub features: CommunityFeatureFlags,
+    pub addon_list: Vec<CommunityAddOnConfigId>,
```
Comment (frol) [inline src/community/mod.rs]: Let's use `Vec<CommunityAddOnConfig>` as in order to render the community page you will anyway need to pull all the configs (the addons themselves can be cached)

File: src/lib.rs [hunk 8]
```diff
@@ -413,6 +418,144 @@ impl Contract {
...
+        {
+            panic!("Only the admin and moderators can edit add-ons");
+        }
+        self.available_addons.insert(&input.id.clone(), &input);
+    }
+
+    pub fn get_community_addon_configs(
+        &self,
+        handle: CommunityHandle,
+    ) -> Vec<CommunityAddOnConfig> {
```
Comment (frol) [inline src/lib.rs]: This method won't be needed as the list of configs will be part of the Community struct.

File: src/lib.rs [hunk 9]
```diff
@@ -44,6 +44,8 @@ pub struct Contract {
     pub authors: UnorderedMap<AccountId, HashSet<PostId>>,
     pub communities: UnorderedMap<CommunityHandle, Community>,
     pub featured_communities: Vec<FeaturedCommunity>,
+    pub available_addons: UnorderedMap<CommunityAddOnId, CommunityAddOn>,
+    pub addon_configurations: LookupMap<CommunityAddOnConfigId, CommunityAddOnConfig>,
```
Comment (frol) [inline src/lib.rs]: As mentioned above, let's move these to the Community struct. It does not make a lot of sense to reuse addon configurations between communities as it won't be possible to identify who has permission to edit the configurations.

File: src/lib.rs (hunk 3 above)
Comment (frol) [reply src/lib.rs]: We won't need Config ID in the first place, if we just move the list of Configs to the Community struct.

**Review by frol — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 3 above)
Comment (Tguntenaar) [reply src/lib.rs]: > would it be better to use a `LookupMap` for the community plugin configuration data? That way it would be possible to look up the config for the plugin in view, and also not make the community config object potentially very large.

@frol, @petersalomonsen adviced that this would cause the community to become too big in some cases.

File: src/community/mod.rs (hunk 4 above)
Comment (Tguntenaar) [reply src/community/mod.rs]: `name` is used for display different tab names of the same addon with different parameters. For instance Wiki1, Wiki2

Comment (frol) [reply src/community/mod.rs]: @Tguntenaar Leave a comment in the code and consider renaming to `display_name`

File: src/lib.rs (hunk 9 above)
Comment (petersalomonsen) [reply src/lib.rs]: I don't think the intention is to reuse addon configurations between communities. The reason for a lookup map is to not having to load the config for all addons in a community. 

But when you mention that, what probably should be done is to make sure that a `communityAddonId` and `communityAddonConfigId` is unique to the community.

**Comment by elliotBraem** [comment]
@frol, @petersalomonsen, Based on the comments + discussion above, here's a diagram of the contract changes that @Tguntenaar and I have decided on implementing, and the way it is reflected on the UI:

image

**Changes**
1. We've renamed "name" to "display_name".
2. Community Struct now holds "addons" and "configs". @frol 
* Addons is a Vector<CommunityAddon> and largely dictates the tabs themselves (the order, enable/disable, display_name... these can be configured from the Community page since it dictates its interaction with it. It holds a reference to a CommunityConfig via config_id.
* Configs is a LookupMap<(CommunityHandle, CommunityConfigId), CommunityConfig>, which is accessed when a tab is selected via community.configs[config_id] @petersalomonsen 
3. We pass the addon_id and the config to the "Addon" that renders, which can either be of TemplateType "viewer" or "configurator". The addon only has accessed to the parameters and any information that the Addon itself provides. @ori-near , see diagram for how the order, tab name, icon, etc. of an addon can be modified via Community, but the parameters for something like the Kanban board, would be configured within the tab itself through the Addon's configurator view. 

The goal was to keep it as modular as possible and with a separation of duties; This way, I think we could also implement permissions control for individual addon configs? And maybe more, I think  this pattern could be generalized and used for more structures similar to communities. @carina-akaia @ailisp @jaswinder6991

Here you can test a dummy example :

**[Check out the Screams and Whispers of Vlad's Haunted Hack House!](https://test.near.org/devhub.efiz.testnet/widget/DevHub.App?page=community&handle=vlads-haunted-hack-house&tab=Screams%20and%20Whispers)**

File: src/lib.rs (hunk 9 above)
Comment (frol) [reply src/lib.rs]: I don't expect configs to be bigger than 1kb each and not more than 10 addons per community, so we are talking about 10kb. It is nothing from CPU time, calling a host function to read a single key is more expensive than deserializing 10kb. This "optimization" complicates things, and potentially makes it worse in the end.

Comment (frol) [reply src/lib.rs]: If you still want to go with storing CommunityAddOnConfig as individual keys, move `LookupMap` into `Community` struct and use the `StorageKey::CommunityAddonConfig(CommunityHandle)` as the storage key prefix.

Comment (frol) [reply src/lib.rs]: On the second thought, given that everything is already implemented, I vote for keeping it as is and address the collision issue - you can use a tuple as a key: `LookupMap<(CommunityHandle, CommunityAddOnConfigId), CommunityAddOnConfig>`

**Comment by petersalomonsen** [comment]
I think this looks great!

Just wondering when referencing `community.configs[addon_id]`. Is then `addon_id` unique for the community? We don't want other communities to get the same config if using the same addon as another community. Or have I missed something?

**Comment by elliotBraem** [comment]
Lol sorry, there are a lot of moving parts and it's hard to keep the naming straight! You are correct, in the diagram, this should not be addon_id, but rather "config_id", where config_id is unique to the community via the Tuple: LookupMap<(CommunityHandle, CommunityAddOnConfigId), CommunityAddOnConfig> as suggested by @frol 

I have updated the original post

File: src/community/mod.rs [hunk 10]
```diff
@@ -47,6 +64,104 @@ pub struct WikiPage {
...
+#[serde(crate = "near_sdk::serde")]
+pub struct CommunityAddOn {
+    pub addon_id: AddOnId,
+    pub display_name: String,
+    pub icon: String,
+    pub enabled: bool,
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct AddOn {
```
Comment (Tguntenaar) [inline src/community/mod.rs:62]: If I add `#[serde(crate = "near_sdk::serde")]` to the AddOn struct and derive Serialize. It doesn't satisfy the requirements fo the Serialize trait because of line 88. (The LookupMap)

Comment (frol) [reply src/community/mod.rs:62]: 1. LookupMap is not serializable as you cannot fetch the values from the storage unless you know the key you used to store it there in the first place.
2. It is expensive to fetch all elements of near-sdk collections - those are designed to store thousands records and thus iterating over the whole collection is a foot gun (you would run out of gas), so thanks to the fact that these methods are not implemented, it stopped you here.

Why do you need a LookupMap for 2 elements? This is some overengineering.

File: src/lib.rs (hunk 3 above)
Comment (frol) [reply src/lib.rs]: Just to bring more context from the thread below: https://github.com/near/neardevhub-contract/pull/61#discussion_r1360615452

TL;DR: I don't think it would be too big, and most probably this would use less gas, but I vote for moving fast here and don't hold it back any longer. This already took too much time, really.

File: src/community/mod.rs [hunk 11]
```diff
@@ -110,6 +177,18 @@ impl Community {
         }
     }
 
+    pub fn add_addon(&mut self, addon_config: CommunityAddOn) {
+        self.addons.push(addon_config);
+    }
+
+    pub fn remove_addon(&mut self, addon2: CommunityAddOn) {
```
Comment (frol) [inline src/community/mod.rs]: `addon2`... srsly? 😭

File: src/community/mod.rs [hunk 12]
```diff
@@ -67,6 +132,8 @@ pub struct Community {
     pub wiki1: Option<WikiPage>,
     pub wiki2: Option<WikiPage>,
     pub features: CommunityFeatureFlags,
+    pub addons: Vec<CommunityAddOn>,
+    pub configs: Vec<AddOnConfig>,
```
Comment (frol) [inline src/community/mod.rs]: There is no sense in splitting up these two. `CommunityAddOnConfig` should be, basically, a many-to-many connection between Community and AddOn entities.

File: src/lib.rs [hunk 13]
```diff
@@ -413,6 +417,102 @@ impl Contract {
             .collect()
     }
 
+    pub fn get_addon(&self, id: AddOnId) -> Option<AddOn> {
```
Comment (frol) [inline src/lib.rs:419]: Align the naming

```suggestion
    pub fn get_available_addon(&self, id: AddOnId) -> Option<AddOn> {
```

File: src/lib.rs [hunk 14]
```diff
@@ -413,6 +417,102 @@ impl Contract {
...
+        self.available_addons.get(&id)
+    }
+
+    pub fn get_available_addons(&self) -> Vec<AddOn> {
+        self.available_addons.iter().map(|(_id, add_on)| add_on).collect()
+    }
+
+    // Only the contract admin and DevHub moderators
+    pub fn create_new_addon(&mut self, addon: AddOn) {
```
Comment (frol) [inline src/lib.rs]:
```suggestion
    pub fn create_addon(&mut self, addon: AddOn) {
```

File: src/lib.rs [hunk 15]
```diff
@@ -413,6 +417,102 @@ impl Contract {
...
+        }
+        let addon = self
+            .get_addon(id.clone())
+            .expect(&format!("Add-on with id `{}` does not exist", id))
+            .clone();
+
+        self.available_addons.remove(&addon.id);
+    }
+
+    pub fn update_addon(&mut self, input: AddOn) {
```
Comment (frol) [inline src/lib.rs]: `input` is really not a great name, at least, let's use `addon`:

```suggestion
    pub fn update_addon(&mut self, addon: AddOn) {
```

File: src/lib.rs [hunk 16]
```diff
@@ -413,6 +417,102 @@ impl Contract {
...
+    }
+
+    pub fn get_community_addons(&self, handle: CommunityHandle) -> Vec<CommunityAddOn> {
+        let community = self
+            .get_community(handle.clone())
+            .expect(format!("Community not found with handle `{}`", handle).as_str());
+        community.addons
+    }
+
+    pub fn set_community_addons(&mut self, handle: CommunityHandle, addons: Vec<CommunityAddOn>) {
```
Comment (frol) [inline src/lib.rs:473]: I totally miss the point of community addons vs community addon configs

**Review by frol — CHANGES_REQUESTED** [review]

File: src/community/mod.rs (hunk 10 above)
Comment (Tguntenaar) [reply src/community/mod.rs:62]: I already thought so, I had reverted it back to a Vec yesterday.

File: src/community/mod.rs (hunk 12 above)
Comment (Tguntenaar) [reply src/community/mod.rs]: removed it

**Comment by Tguntenaar** [comment]
Thank you everyone for the comments and review points. I deployed the latest version on [thomaselliot.testnet](https://explorer.testnet.near.org/transactions/EyBTUrVtb3CtiViMCrHmETqEjJrHvi7keHZ322cV6kPK)

File: src/lib.rs [hunk 17]
```diff
@@ -413,6 +416,82 @@ impl Contract {
...
+        self.update_community(handle, community);
+    }
+
+    // To add or update parameters set by the configurator widget
+    pub fn set_community_addon(self, handle: CommunityHandle, addon: CommunityAddOn) {
+        let mut community = self
+            .get_community(handle.clone())
+            .expect(format!("Community not found with handle `{}`", handle).as_str());
+        if let Some(index) =
+            community.addons.iter().position(|current| current.addon_id == addon.addon_id)
```
Comment (frol) [inline src/lib.rs]: Such implementation limits usage of several same widgets (e.g. Wiki widget) for a single community. You can still hack it around by setting the whole list of addons, but I still wanted to mention this bug.

**Review by frol — COMMENTED** [review]
Deleted comment

File: src/post/mod.rs [hunk 18]
```diff
@@ -53,6 +53,9 @@ pub enum StorageKey {
     AuthorToAuthorPosts,
     AuthorPosts(CryptoHash),
     Communities,
+    AddOns,
+    AddOnsConfig,
+    TemplateType,
```
Comment (frol) [inline src/post/mod.rs]: These should be removed now

**Review by frol — APPROVED** [review]
Overall, it looks good, and ready to merge after the last minor comments are addressed and merge conflict is resolved

File: src/post/mod.rs (hunk 18 above)
Comment (frol) [inline src/post/mod.rs]:
```suggestion
```

File: src/lib.rs (hunk 17 above)
Comment (Tguntenaar) [reply src/lib.rs]: I added an id to the CommunityAddon to be able to differentiate them

File: src/post/mod.rs (hunk 18 above)
Comment (Tguntenaar) [reply src/post/mod.rs]: Removed

**Comment by Tguntenaar** [comment]
@frol it's finally ready to merge!

File: src/lib.rs [hunk 19]
```diff
@@ -594,4 +707,156 @@ mod tests {
...
+    //     let addon = CommunityAddOn {
+    //         addon_id: "CommunityAddOnId".to_string(),
+    //         display_name: "GitHub".to_string(),
+    //         enabled: true,
+    //         parameters: "".to_string(),
+    //     };
+    //     contract.set_community_addon(community_handle.to_owned(), addon.clone());
+    //     let retreived_addons = contract.get_community_addons(community_handle.to_string());
+    //     assert_eq!(addon.addon_id, retreived_addons[0].addon_id);
+    // }
```
Comment (frol) [inline src/lib.rs]: Why is this commented out?

File: src/lib.rs [hunk 20]
```diff
@@ -410,6 +416,82 @@ impl Contract {
...
+        self.update_community(handle, community);
+    }
+
+    // To add or update parameters set by the configurator widget
+    pub fn set_community_addon(self, handle: CommunityHandle, community_addon: CommunityAddOn) {
+        let mut community = self
+            .get_community(handle.clone())
+            .expect(format!("Community not found with handle `{}`", handle).as_str());
+        if let Some(index) =
+            community.addons.iter().position(|current| current.id == community_addon.id)
```
Comment (frol) [inline src/lib.rs]: You can use [`find`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.find) instead of `position`, so you don't need to `[index]` then.

File: src/lib.rs [hunk 21]
```diff
@@ -410,6 +416,82 @@ impl Contract {
...
+        {
+            community.addons[index] = community_addon;
+        } else {
+            community.addons.push(community_addon);
+        }
+    }
```
Comment (frol) [inline src/lib.rs:498]: I believe you need to call `update_community` to persist the changes:

```suggestion
        self.update_community(handle, community);
    }
```

File: src/lib.rs [hunk 22]
```diff
@@ -433,7 +515,10 @@ impl Contract {
         if target_community.handle == community.handle {
             self.communities.insert(&target_community.handle, &community);
         } else {
-            require!(self.communities.get(&community.handle).is_none(), "Community handle `{community.handle}` is already taken");
+            require!(
+                self.communities.get(&community.handle).is_none(),
+                "Community handle `{community.handle}` is already taken"
```
Comment (frol) [inline src/lib.rs]: I believe formatting strings do not work in `require!`, please double-check that it actually substitutes the `{community.handle}` with the value, and not just leaving it as "{community.handle}" text.

Comment (petersalomonsen) [reply src/lib.rs]: better add a unit test to confirm this?

Comment (frol) [reply src/lib.rs]: @petersalomonsen Sure, that is even better

Comment (Tguntenaar) [reply src/lib.rs]: You're right it didn't work indeed, I removed the template from the string.

File: src/lib.rs (hunk 20 above)
Comment (Tguntenaar) [reply src/lib.rs]: Done

**Comment by Tguntenaar** [comment]
@frol one more try! :)

**Review by frol — APPROVED** [review]
Looks good to me now
